src/ast.rs

```diff
@@ -0,0 +1,162 @@
+            })
+    }
+}
+
+impl<'s> Expression<Spanned<'s>> {
+    /// Parses the given `input` as an [`Expression`].
+    ///
+    /// # Errors
+    ///
+    /// See [`parse::Error`] for details.
+    pub fn parse<I: AsRef<str>>(
+        input: &'s I,
+    ) -> Result<Self, parse::Error<Spanned<'s>>> {
+        Self::try_from(input.as_ref())
+    }
+}
+
+/// Single entry of a [`cucumber-expression`][3].
+///
+/// See [`parse::single_expression()`] for the detailed grammar and examples.
+///
+/// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum SingleExpression<Input> {
+    /// [`alternation`][3] expression.
+    ///
+    /// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+    Alternation(Alternation<Input>),
+
+    /// [`optional`][3] expression.
+    ///
+    /// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+    Optional(Optional<Input>),
+
+    /// [`parameter`][3] expression.
+    ///
+    /// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+    Parameter(Parameter<Input>),
+
+    /// Text without whitespaces.
+    Text(Input),
+
+    /// Whitespaces are treated as a special case to avoid lookaheads and
+    /// lookbehinds described in the [architecture][1]. This allows parsing to
+    /// have `O(n)` complexity.
+    ///
+    /// [1]: https://tinyurl.com/cucumber-expr-spec
+    Whitespace,
+}
+
+/// Allows to match one of [`SingleAlternation`]s.
+///
+/// See [`parse::alternation()`] for detailed syntax and examples.
+#[derive(AsRef, Clone, Debug, Deref, DerefMut, Eq, PartialEq)]
+pub struct Alternation<Input>(pub Vec<SingleAlternation<Input>>);
+
+/// Building block an [`Alternation`].
+pub type SingleAlternation<Input> = Vec<Alternative<Input>>;
+
+impl<Input: InputLength> Alternation<Input> {
+    /// Returns length of capture from `Input`.
+    pub(crate) fn span_len(&self) -> usize {
+        self.0
+            .iter()
+            .flatten()
+            .map(|alt| match alt {
+                Alternative::Text(t) => t.input_len(),
+                Alternative::Optional(opt) => opt.input_len() + 2,
+            })
+            .sum::<usize>()
+            + self.len()
+            - 1
+    }
+
+    /// Indicates whether one of [`SingleAlternation`]s consists only from
+    /// [`Optional`]s.
+    pub(crate) fn contains_only_optional(&self) -> bool {
+        for single_alt in &**self {
+            if single_alt
+                .iter()
+                .all(|alt| matches!(alt, Alternative::Optional(_)))
+            {
+                return true;
+            }
+        }
+        false
+    }
+}
+
+/// [`alternative`][3] expression.
+///
+/// See [`parse::alternative()`] for the detailed grammar and examples.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub enum Alternative<Input> {
+    /// [`optional`][3] expression.
+    ///
+    /// [3]: https://tinyurl.com/cucumber-expr-spec#grammar
+    Optional(Optional<Input>),
+
+    /// Text.
+    Text(Input),
```

Comment:
This deviates from the [original grammar](https://github.com/cucumber/cucumber-expressions/blob/main/ARCHITECTURE.md#grammar) by missing the `parameter` variant.

Need to discuss this on voice.